public/app/core/components/TraceToLogs/TagMappingInput.tsx

```diff
@@ -25,14 +25,14 @@ export const TagMappingInput = ({ values, onChange, id }: Props) => {
               placeholder={'Tag name'}
               value={value.key}
               onChange={(e) => {
-                onChange(
-                  values.map((v, i) => {
-                    if (i === idx) {
-                      v.key = String(e);
-                    }
-                    return v;
-                  })
-                );
+                const vals = [...values];
+                vals.map((v, i) => {
+                  if (i === idx) {
+                    v.key = String(e);
+                  }
+                  return v;
+                });
+                onChange(vals);
```

Comment:
This does not seem like it should change anything. `values.map` should have already made a copy of the array so I don't think the array mutation was a problem. Could be that modification of the value was the problem but in that case this new code does the same thing. Also before it the `map` result was actually used. Now this actually uses `map` only to do a side effect modification of the `vals` array which is not a good use of `map`.

Reply:
Yeah thanks for that! I've made changes that deepClone the values and also how the tag is updated within the values array.

Reply:
I don't think we should be cloning the array here. Not a big deal in this part of code but overall as a strategy is not good.

1. Deep copy of the whole array is probably more costly than just copying the array and changing single value.
2. Semantically this is not what needs to happen and will potentially crate more rerenders in react that are not necessary. Anything that relies on an item from the array would see a new object even though the object values did not actually change so rerender isn't necessary.

Simplest thing that seems like should help here is:
```javascript
values.map((v, i) => {
  if (i === idx) {
    return { ...v, key: String(e)}
  }
  return v;
}
```

This maps the array, which creates a copy of the array. This makes sense because anything using the while array should see that something in it changed. It just passes the value if nothing needs to change, so anything requiring that value will see the same object ref and don't need to rerender. The value that changed was recreated as a new object so it should trigger a rerender and should not be seen as updating a read only value.